modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/HeadlessDatasetDisplayTag.java

```diff
@@ -0,0 +1,408 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.taglib.clay.servlet.taglib;
+
+import com.liferay.frontend.js.loader.modules.extender.npm.NPMResolver;
+import com.liferay.frontend.taglib.clay.data.set.ClayDataSetActionDropdownItem;
+import com.liferay.frontend.taglib.clay.data.set.ClayDataSetDisplayViewSerializer;
+import com.liferay.frontend.taglib.clay.data.set.ClayDataSetFilterSerializer;
+import com.liferay.frontend.taglib.clay.data.set.model.ClayPaginationEntry;
+import com.liferay.frontend.taglib.clay.internal.js.loader.modules.extender.npm.NPMResolverProvider;
+import com.liferay.frontend.taglib.clay.internal.servlet.ServletContextUtil;
+import com.liferay.frontend.taglib.clay.servlet.taglib.util.CreationMenu;
+import com.liferay.frontend.taglib.clay.servlet.taglib.util.DropdownItem;
+import com.liferay.portal.kernel.dao.search.SearchContainer;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.util.PortalUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.util.PropsValues;
+import com.liferay.taglib.util.IncludeTag;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+
+import javax.portlet.PortletURL;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.jsp.JspException;
+import javax.servlet.jsp.PageContext;
+
+/**
+ * @author Marco Leo
+ * @author Alessio Antonio Rendina
+ */
+public class HeadlessDatasetDisplayTag extends IncludeTag {
```

Comment:
Just for future improvements, since we won't be using jsps, it'd be better to drop `IncludeTag` from the hierarchy as it makes everything needlessly slow.

We could make this extend the new `BaseContainerTag` and then hydrate it using the existing mechanisms so it behaves exactly as the other `clay:*` tags.

That way, we could even leverage SSR if we ever come to that

Reply:
I stand corrected... we use `jsp`s but I don't really know why 😂